File: zig/thoth-ipc/src/transport/notify.zig
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya and thoth-ipc contributors
//
// Layer 1 notify readiness (ABI §8, macOS libnotify backend). A sender posts on
// enqueue; an async receiver registers a readiness fd and wakes on the post. The
// service key is byte-exact across languages so a Zig sender wakes a C++/Rust/
// Swift async receiver and vice versa:
//   key = "thoth.ntf." + 16-hex(fnv1a_64("{prefix}__THOTH_SHM__NOTIFY__{name}"))
// Golden: ("","xchan") -> 098e889ce378ae04 (unit-tested in shmname.zig).

const std = @import("std");
const shmname = @import("../platform/shmname.zig");

const NOTIFY_STATUS_OK: u32 = 0;
const POLLIN: i16 = 0x0001;
const O_NONBLOCK: c_int = 0x0004; // Darwin

// libnotify (libSystem) — private-but-stable notification service.
extern "c" fn notify_post(name: [*:0]const u8) u32;
extern "c" fn notify_register_file_descriptor(name: [*:0]const u8, notify_fd: *c_int, flags: c_int, out_token: *c_int) u32;
extern "c" fn notify_cancel(token: c_int) u32;

/// Build the libnotify service key for (prefix, name) into `buf`, NUL-terminated.
fn serviceKey(buf: []u8, prefix: []const u8, name: []const u8) [:0]const u8 {
    var sbuf: [256]u8 = undefined;
    const s = std.fmt.bufPrint(&sbuf, "{s}__THOTH_SHM__NOTIFY__{s}", .{ prefix, name }) catch unreachable;
    const hash = shmname.fnv1a64(s);
    const key = std.fmt.bufPrint(buf, "thoth.ntf.{x:0>16}\x00", .{hash}) catch unreachable;
    return key[0 .. key.len - 1 :0];
}

/// Post the readiness signal (multicast — one post wakes every registered
/// reader). A no-op if nobody is listening, so it is cheap to call on every send.
pub fn post(prefix: []const u8, name: []const u8) void {
    var buf: [64]u8 = undefined;
    _ = notify_post(serviceKey(&buf, prefix, name).ptr);
}

/// A receiver's readiness fd — signalled (a 4-byte token per post) whenever any
/// language's sender posts. Level-triggered: drain the queued tokens after a wake.
pub const Sink = struct {
    fd: c_int = -1,
    token: c_int = 0,
    valid: bool = false,

    pub fn open(prefix: []const u8, name: []const u8) Sink {
        var buf: [64]u8 = undefined;
        const key = serviceKey(&buf, prefix, name);
        var self = Sink{};
        if (notify_register_file_descriptor(key.ptr, &self.fd, 0, &self.token) != NOTIFY_STATUS_OK) return self;
        // Non-blocking so drain() can read tokens until they run out.
        const fl = std.c.fcntl(self.fd, std.c.F.GETFL, @as(c_int, 0));
        _ = std.c.fcntl(self.fd, std.c.F.SETFL, fl | O_NONBLOCK);
        self.valid = true;
        return self;
    }

    /// Block until the fd signals or `timeout_ms` elapses. Returns true on signal.
    pub fn wait(self: *const Sink, timeout_ms: i32) bool {
        var pfd = std.c.pollfd{ .fd = self.fd, .events = POLLIN, .revents = 0 };
        const n = std.c.poll(@ptrCast(&pfd), 1, timeout_ms);
        return n > 0 and (pfd.revents & POLLIN) != 0;
    }

    /// Drain queued readiness tokens (4-byte ints) until the fd would block.
    pub fn drain(self: *const Sink) void {
        var scratch: [64]u8 = undefined;
        while (std.c.read(self.fd, &scratch, scratch.len) > 0) {}
    }

    pub fn close(self: *Sink) void {
        if (!self.valid) return;
        _ = notify_cancel(self.token);
        _ = std.c.close(self.fd);
        self.valid = false;
    }
};

test "service key matches the golden notify hash" {
    var buf: [64]u8 = undefined;
    const key = serviceKey(&buf, "", "xchan");
    try std.testing.expectEqualStrings("thoth.ntf.098e889ce378ae04", key);
}

File: swift/thoth-ipc/Sources/ThothIPC/Transport/Notify.swift
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya and thoth-ipc contributors
//
// Layer 1 of the optional async-receive work, Swift side. **Byte-exact with
// cpp/thoth-ipc/src/thoth-ipc/notify.h and rust/thoth-ipc/src/notify.rs** so a Swift
// `send()` wakes a C++ `async_recv` / Rust `AsyncRoute` (and vice versa). Swift
// is macOS-only, so this is the Darwin libnotify backend: `notify_post(key)`
// wakes an fd from `notify_register_file_descriptor(key, ...)` in ANY process,
// multicast (one post wakes every registered reader), one key per channel.
//
// The key is `"ipc.ntf." + 16-hex FNV-1a-64 of "{prefix}__IPC_SHM__NOTIFY__{name}"`.

import Darwin
import ThothIPCShim

/// 16-hex FNV-1a-64 of make_public_abi_prefix(prefix, "NOTIFY__", name). Byte-exact with C++/Rust.
func notifyHash(_ prefix: String, _ name: String) -> String {
    toHex16(fnv1a64("\(fullPrefix(prefix))NOTIFY__\(name)".utf8))
}

/// libnotify service key for a channel (multicast; one per channel).
func notifyKey(_ prefix: String, _ name: String) -> String {
    "ipc.ntf.\(notifyHash(prefix, name))"
}

/// Writer side: post the channel's key; libnotify multicasts to all readers.
final class NotifySource {
    private var key: String?

    func signal(_ prefix: String, _ name: String) {
        if key == nil { key = notifyKey(prefix, name) }
        key!.withCString { _ = notify_post($0) }
    }
}

/// Reader side: an fd libnotify writes a token to on every matching post.
final class NotifySink {
    private(set) var fd: Int32 = -1
    private var token: Int32 = -1

    var valid: Bool { fd != -1 }

    /// Register the readiness fd for this channel. Idempotent.
    @discardableResult
    func open(_ prefix: String, _ name: String) -> Bool {
        if fd != -1 { return true }
        var f: Int32 = -1
        var t: Int32 = -1
        let status = notifyKey(prefix, name).withCString {
            notify_register_file_descriptor($0, &f, 0, &t)
        }
        guard status == UInt32(NOTIFY_STATUS_OK) else { return false }
        // Non-blocking so drain() never stalls; cloexec for fd hygiene.
        let fl = fcntl(f, F_GETFL, 0)
        if fl != -1 { _ = fcntl(f, F_SETFL, fl | O_NONBLOCK) }
        _ = fcntl(f, F_SETFD, FD_CLOEXEC)
        fd = f
        token = t
        return true
    }

    /// Consume pending token ints after the fd signalled readable (level-triggered).
    func drain() {
        guard fd != -1 else { return }
        var tok: Int32 = 0
        while read(fd, &tok, MemoryLayout<Int32>.size) > 0 {}
    }

    func close() {
        // notify_cancel closes the fd once its last token is cancelled.
        if token != -1 { notify_cancel(token); token = -1 }
        fd = -1
    }

    deinit { close() }
}
